File: swift/LeetCodeProblems/LeetCodeProblemsTests/LC75_Tests/LC75_Array_Tests.swift
//
//  LC75_Array_Tests.swift
//  LeetCodeProblemsTests
//
//  Created by Aynur Nasybullin on 04.09.2023.
//

import XCTest
@testable import LeetCodeProblems

final class LC75_Array_Tests: XCTestCase {
    // MARK: Problem 443. String Compression
    lazy var testsData_problem_443: [(chars: [Character], result: Int, resultChars: [Character])] = {
        var testsData = [(chars: [Character], result: Int, resultChars: [Character])]()
        
        testsData.append((chars: ["a", "a", "b", "b", "c", "c", "c"],
                          result: 6,
                          resultChars: ["a", "2", "b", "2", "c", "3"]))
        testsData.append((chars: ["a"],
                          result: 1,
                          resultChars: ["a"]))
        testsData.append((chars: ["a", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b"],
                          result: 4,
                          resultChars: ["a", "b", "1", "2"]))
        
        return testsData
    }()
    
    func tests_problem_443_compress() throws {
        for data in testsData_problem_443 {
            var chars = data.chars
            let actual = LC75_Array.problem_443_compress(&chars)
            
            XCTAssertEqual(actual, data.result)
            XCTAssertEqual(chars, data.resultChars)
        }
    }
    
    
    // MARK: Problem 334. Increasing Triplet Subsequence
    lazy var testsData_problem_334: [(nums: [Int], result: Bool)] = {
        var testsData = [(nums: [Int], result: Bool)]()
        
        testsData.append((nums: [1, 2, 3, 4, 5], result: true))
        testsData.append((nums: [5, 4, 3, 2, 1], result: false))
        testsData.append((nums: [2, 1, 5,0,4,6], result: true))
        
        return testsData
    }()
    
    func tests_problem_334_increasingTriplet() throws {
        for data in testsData_problem_334 {
            let actual = LC75_Array.problem_334_increasingTriplet(data.nums)
            XCTAssertEqual(actual, data.result)
        }
    }
    
    
    // MARK: Problem 238. Product of Array Except Self
    lazy var testsData_problem_238: [(nums: [Int], result: [Int])] = {
        var testsData = [(nums: [Int], result: [Int])]()
        
        testsData.append((nums: [1, 2, 3, 4], result: [24, 12, 8, 6]))
        testsData.append((nums: [-1, 1, 0, -3, 3], result: [0, 0, 9, 0, 0]))
        
        return testsData
    }()
    
    func tests_problem_238_productExceptSelf() throws {
        for data in testsData_problem_238 {
            let actual = LC75_Array.problem_238_productExceptSelf(data.nums)
            XCTAssertEqual(actual, data.result)
        }
    }
    
    
    // MARK: Problem 151. Reverse Words in a String
    lazy var testsData_problem_151: [(s: String, result: String)] = {
        var testsData = [(s: String, result: String)]()
        
        testsData.append((s: "the sky is blue", result: "blue is sky the"))
        testsData.append((s: "  hello world  ", result: "world hello"))
        testsData.append((s: "a good   example", result: "example good a"))
        
        return testsData
    }()
    
    func tests_problem_151_reverseWords() throws {
        for data in testsData_problem_151 {
            let actual = LC75_Array.problem_151_reverseWords(data.s)
            XCTAssertEqual(actual, data.result, "s: \(data.s), result: \(data.result)")
        }
    }
    
    
    // MARK: Problem 345. Reverse Vowels of a String
    lazy var testsData_problem_345: [(s: String, result: String)] = {
        var testsData = [(s: String, result: String)]()
        
        testsData.append((s: "hello", result: "holle"))
        testsData.append((s: "leetcode", result: "leotcede"))
        testsData.append((s: "aA", result: "Aa"))
        
        return testsData
    }()
    
    func tests_problem_345_reverseVowels() throws {
        for data in testsData_problem_345 {
            let actual = LC75_Array.problem_345_reverseVowels(data.s)
            XCTAssertEqual(actual, data.result, "s: \(data.s), result: \(data.result)")
        }
    }
    
    
    // MARK: Problem 605. Can Place Flowers
    lazy var testsData_problem_605: [(flowerbed: [Int], n: Int, result: Bool)] = {
        var testsData = [(flowerbed: [Int], n: Int, result: Bool)]()
        
        testsData.append((flowerbed: [0, 0, 0], n: 2, result: true))
        testsData.append((flowerbed: [0, 1, 0], n: 1, result: false))
        testsData.append((flowerbed: [1, 0, 0, 0, 1], n: 1, result: true))
        testsData.append((flowerbed: [1, 0, 0, 0, 1], n: 2, result: false))
        testsData.append((flowerbed: [0, 0, 1, 0, 1], n: 1, result: true))
        testsData.append((flowerbed: [0], n: 1, result: true))
        
        return testsData
    }()
    
    func tests_problem_605_canPlaceFlowers() throws {
        for data in testsData_problem_605 {
            let actual = LC75_Array.problem_605_canPlaceFlowers(data.flowerbed, data.n)
            XCTAssertEqual(actual, data.result, "flowerbed: \(data.flowerbed), n: \(data.n)")
        }
    }
    
    
    // MARK: Problem 1431. Kids With the Greatest Number of Candies
    lazy var testsData_problem_1431: [(candies: [Int], extraCandies: Int, result: [Bool])] = {
        var testsData = [(candies: [Int], extraCandies: Int, result: [Bool])]()
        
        testsData.append((candies: [2, 3, 5, 1, 3], extraCandies: 3, result: [true, true, true, false, true]))
        testsData.append((candies: [4, 2, 1, 1, 2], extraCandies: 1, result: [true, false, false, false, false]))
        testsData.append((candies: [12, 1, 12], extraCandies: 10, result: [true, false, true]))
        
        return testsData
    }()
    
    func tests_problem_1431_kidsWithCandies() throws {
        for data in testsData_problem_1431 {
            let actual = LC75_Array.problem_1431_kidsWithCandies(data.candies, data.extraCandies)
            XCTAssertEqual(actual, data.result)
        }
    }
    
    
    // MARK: Problem 1071. Merge Strings Alternately
    lazy var testsData_problem_1071: [(str1: String, str2: String, result: String)] = {
        var testsData = [(str1: String, str2: String, result: String)]()
        
        testsData.append((str1: "ABCABC", str2: "ABC", result: "ABC"))
        testsData.append((str1: "ABABAB", str2: "ABAB", result: "AB"))
        testsData.append((str1: "LEET", str2: "CODE", result: ""))
        testsData.append((str1: "TAUXXTAUXXTAUXXTAUXXTAUXX", str2: "TAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXXTAUXX", result: "TAUXX"))
        testsData.append((str1: "AAAAAAAAA", str2: "AAACCC", result: ""))
        
        return testsData
    }()
    
    func tests_problem_1071_gcdOfStrings() throws {
        for data in testsData_problem_1071 {
            let actual = LC75_Array.problem_1071_gcdOfStrings(data.str1, data.str2)
            XCTAssertEqual(actual, data.result)
        }
    }

    
    // MARK: Problem 1768. Merge Strings Alternately
    lazy var testsData_problem_1768: [(word1: String, word2: String, result: String)] = {
        var testsData = [(word1: String, word2: String, result: String)]()
        
        testsData.append((word1: "abc", word2: "pqr", result: "apbqcr"))
        testsData.append((word1: "ab", word2: "pqrs", result: "apbqrs"))
        testsData.append((word1: "abcd", word2: "pq", result: "apbqcd"))
        
        return testsData
    }()
    
    func tests_problem_1768_mergeAlternately() throws {
        for data in testsData_problem_1768 {
            let actual = LC75_Array.problem_1768_mergeAlternately(data.word1, data.word2)
            XCTAssertEqual(actual, data.result)
        }
    }
}

File: cpp/leetCode75/lc75_01_array/lc75_array_tests.cpp
#include <tuple>

#include <gtest/gtest.h>

#include "lc75_array.h"

void printVector(std::vector<bool> items) {
    for (int item : items) {
        std::cout << item << " ";
    }
}


#pragma region Problem 1768. Merge Strings Alternately
class Problem_1768_Tests : public ::testing::TestWithParam<std::tuple<std::string, std::string, std::string>> { };

std::tuple<std::string, std::string, std::string> testCases_problem_1768[] {
    std::make_tuple("abc", "pqr", "apbqcr"),
    std::make_tuple("ab", "pqrs", "apbqrs"),
    std::make_tuple("abcd", "pq", "apbqcd")
};

TEST_P(Problem_1768_Tests, BasicTests) {
    std::string word1, word2, result;
    std::tie(word1, word2, result) = GetParam();

    auto actual = problem_1768_mergeAlternately(word1, word2);
    EXPECT_EQ(actual, result);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_1768_Tests, ::testing::ValuesIn(testCases_problem_1768));
#pragma endregion


#pragma region Problem 1071. Greatest Common Divisor of Strings
class Problem_1071_Tests : public ::testing::TestWithParam<std::tuple<std::string, std::string, std::string>> { };

std::tuple<std::string, std::string, std::string> testCases_problem_1071[] {
    std::make_tuple("ABCABC", "ABC", "ABC"),
    std::make_tuple("ABABAB", "ABAB", "AB"),
    std::make_tuple("LEET", "CODE", "")
};

TEST_P(Problem_1071_Tests, BasicTests) {
    std::string str1, str2, result;
    std::tie(str1, str2, result) = GetParam();

    auto actual = problem_1071_gcdOfStrings(str1, str2);
    EXPECT_EQ(actual, result);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_1071_Tests, ::testing::ValuesIn(testCases_problem_1071));
#pragma endregion


#pragma region Problem 1431. Kids With the Greatest Number of Candies
class Problem_1431_Tests : public ::testing::TestWithParam<std::tuple<std::vector<int>, int, std::vector<bool>>> { };

std::vector<int> candies1 = { 2, 3, 5, 1, 3 };
std::vector<int> candies2 = { 4, 2, 1, 1, 2 };
std::vector<int> candies3 = { 2, 1, 12 };

std::vector<bool> result1 = { true, true, true, false, true };
std::vector<bool> result2 = { true, false, false, false, false };
std::vector<bool> result3 = { true, false, true };

std::tuple<std::vector<int>, int, std::vector<bool>> testsCases_problem_1431[] {
    std::make_tuple(candies1, 3, result1),
    std::make_tuple(candies2, 1, result2),
    std::make_tuple(candies3, 10, result3),
};

TEST_P(Problem_1431_Tests, BasicTests) {
    std::vector<int> candies;
    int extraCandies;
    std::vector<bool> result;
    std::tie(candies, extraCandies, result) = GetParam();

    auto actual = problem_1431_kidsWithCandies(candies, extraCandies);
    EXPECT_EQ(std::equal(actual.begin(), actual.end(), result.begin(), result.end()), true);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_1431_Tests, ::testing::ValuesIn(testsCases_problem_1431));
#pragma endregion


#pragma region Problem 345. Reverse Vowels of a String
class Problem_345_Tests : public ::testing::TestWithParam<std::tuple<std::string, std::string>> { };

std::tuple<std::string, std::string> testsCases_problem_345[] {
    std::make_tuple("hello", "holle"),
    std::make_tuple("leetcode", "leotcede"),
};

TEST_P(Problem_345_Tests, BasicTests) {
    std::string s, result;
    std::tie(s, result) = GetParam();

    auto actual = problem_345_reverseVowels(s);
    EXPECT_EQ(actual, result);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_345_Tests, ::testing::ValuesIn(testsCases_problem_345));
#pragma endregion


#pragma region Problem 151. Reverse Words in a String
class Problem_151_Tests : public ::testing::TestWithParam<std::tuple<std::string, std::string>> { };

std::tuple<std::string, std::string> testCases_problem_151[] {
    std::make_tuple("the sky is blue", "blue is sky the"),
    std::make_tuple("  hello world  ", "world hello"),
    std::make_tuple("a good   example", "example good a"),
};

TEST_P(Problem_151_Tests, BasicTests) {
    std::string s, expected;
    std::tie(s, expected) = GetParam();

    auto actual = problem_151_reverseWords(s);
    EXPECT_EQ(actual, expected);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_151_Tests, ::testing::ValuesIn(testCases_problem_151));
#pragma endregion


#pragma region Problem 605. Can Place Flowers
class Problem_605_Tests : public ::testing::TestWithParam<std::tuple<std::vector<int>, int, bool>> { };

std::vector<int> flowerbed_1 = {1, 0, 0, 0, 1};
std::vector<int> flowerbed_2 = {1, 0, 0, 0, 1};
std::vector<int> flowerbed_109 = {1, 0, 0, 0, 0, 1};
std::vector<int> flowerbed_123 = {0, 1, 0};
std::vector<int> flowerbed_112 = {0, 0, 1, 0, 1};

std::tuple<std::vector<int>, int, bool> testsCases_problem_605[] {
    std::make_tuple(flowerbed_1, 1, true),
    std::make_tuple(flowerbed_2, 2, false),
    std::make_tuple(flowerbed_109, 2, false),
    std::make_tuple(flowerbed_123, 1, false),
    std::make_tuple(flowerbed_112, 1, true),
};

TEST_P(Problem_605_Tests, BasicTests) {
    std::vector<int> flowerbed;
    int n;
    bool result;
    std::tie(flowerbed, n, result) = GetParam();

    auto actual = problem_605_canPlaceFlowers(flowerbed, n);
    EXPECT_EQ(actual, result);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_605_Tests, ::testing::ValuesIn(testsCases_problem_605));
#pragma endregion


#pragma region Problem 238. Product of Array Except Self
class Problem_238_Tests : public ::testing::TestWithParam<std::tuple<std::vector<int>, std::vector<int>>> { };

std::vector<int> input_238_1 = { 1, 2, 3, 4 };
std::vector<int> input_238_2 = { -1, 1, 0, -3, 3 };

std::vector<int> output_238_1 = { 24, 12, 8, 6 };
std::vector<int> output_238_2 = { 0, 0, 9, 0, 0 };

std::tuple<std::vector<int>, std::vector<int>> testCases_problem_238[] {
    std::make_tuple(input_238_1, output_238_1),
    std::make_tuple(input_238_2, output_238_2),
};

TEST_P(Problem_238_Tests, BasicTests) {
    std::vector<int> nums, expected;
    std::tie(nums, expected) = GetParam();

    auto actual = problem_238_productExceptSelf(nums);
    // EXPECT_EQ(std::equal(actual.begin(), actual.end(), expected.begin(), expected.end()), true);
    EXPECT_EQ(actual, expected);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_238_Tests, ::testing::ValuesIn(testCases_problem_238));
#pragma endregion


#pragma region Problem 334. Increasing Triplet Subsequence
class Problem_334_Tests : public ::testing::TestWithParam<std::tuple<std::vector<int>, bool>> { };

std::vector<int> input_334_1 = { 1, 2, 3, 4, 5 };
std::vector<int> input_334_2 = { 5, 4, 3, 2, 1 };
std::vector<int> input_334_3 = { 2, 1, 5, 0, 4, 6 };

std::tuple<std::vector<int>, bool> testCases_problem_334[] {
    std::make_tuple(input_334_1, true),
    std::make_tuple(input_334_2, false),
    std::make_tuple(input_334_3, true),
};

TEST_P(Problem_334_Tests, BasicTests) {
    std::vector<int> nums;
    bool expected;
    std::tie(nums, expected) = GetParam();

    auto actual = problem_334_increasingTriplet(nums);
    EXPECT_EQ(actual, expected);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_334_Tests, ::testing::ValuesIn(testCases_problem_334));
#pragma endregion


#pragma region Problem 443. String Compression
class Problem_443_Tests : public ::testing::TestWithParam<std::tuple<std::vector<char>, int, std::vector<char>>> { };

std::vector<char> input_443_1 = { 'a', 'a', 'b', 'b', 'c', 'c', 'c' };
std::vector<char> input_443_2 = { 'a' };
std::vector<char> input_443_3 = { 'a', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b' };

std::vector<char> output_443_1 = { 'a', '2', 'b', '2', 'c', '3' };
std::vector<char> output_443_2 = { 'a' };
std::vector<char> output_443_3 = { 'a', 'b', '1', '2' };

std::tuple<std::vector<char>, int, std::vector<char>> testCases_problem_443[] {
    std::make_tuple(input_443_1, 6, output_443_1),
    std::make_tuple(input_443_2, 1, output_443_2),
    std::make_tuple(input_443_3, 4, output_443_3),
};

TEST_P(Problem_443_Tests, BasicTests) {
    std::vector<char> chars, expectedChars;
    int expected;
    std::tie(chars, expected, expectedChars) = GetParam();

    auto actual = problem_443_compress(chars);
    EXPECT_EQ(actual, expected);
    EXPECT_EQ(chars, expectedChars);
};

INSTANTIATE_TEST_SUITE_P(Default, Problem_443_Tests, ::testing::ValuesIn(testCases_problem_443));
#pragma endregion